app/services/setting_registry.rb

```diff
@@ -1,6 +1,62 @@
 class SettingRegistry
   include Singleton
   include Enumerable
+  extend ScopedSearch::ClassMethods
+
+  class SettingCompleter < ScopedSearch::AutoCompleteBuilder
+    def initialize(registry, definition, query, options)
+      @registry = registry
+      super(definition, query, options)
+    end
+
+    def self.auto_complete(registry, definition, query, options)
+      return [] if (query.nil? || definition.nil? || !definition.respond_to?(:fields))
+
+      new(registry, definition, query, options).build_autocomplete_options
+    end
+
+    def is_query_valid
+      true
+    end
+
+    def complete_value
+      if last_token_is(COMPARISON_OPERATORS)
+        token = tokens[tokens.size - 2]
+        val = ''
+      else
+        token = tokens[tokens.size - 3]
+        val = tokens[tokens.size - 1]
+      end
+
+      field = definition.field_by_name(token)
+      return [] unless field&.complete_value
+      return complete_set(field) if field.set?
+      return complete_key_value(field, token, val) if field.key_field
+
+      special_values = field.special_values.select { |v| v =~ /\A#{val}/ }
+      special_values + complete_value_from_db(field, special_values, val)
+    end
+
+    def complete_value_from_db(field, special_values, val)
+      count = 20 - special_values.count
+      if field.field == :name
+        results = @registry.map { |set| (set.full_name =~ /\s/ ? "\"#{set.full_name.gsub('"', '\"')}\"" : set.full_name) if set.name.include?(val) || set.full_name&.include?(val) }
+      elsif field.field == :description
+        results = []
+      else
+        raise ::Foreman::Exception.new N_('Unsupported completion, only name and description are supported')
+      end
+      results.first(count)
+    end
+  end
+
+  scoped_search :on => :id, :complete_enabled => false, :only_explicit => true, :validator => ScopedSearch::Validators::INTEGER
+  scoped_search on: :name, complete_value: :true, operators: ['~', "="]
+  scoped_search on: :description, complete_value: :true, operators: ['~']
```

Comment:
we have the definitions in the setting model as well for hammer purpose, but we need it in the registry as well because we need to pass the scoped search definition in the auto_complete method (which we are overriding)

Reply:
That's not a problem, just pass the definition from the `Setting` model:
```
    SettingCompleter.auto_complete(self, Setting.scoped_search_definition, query, opts)
```
Otherwise we are creating duplication of search definitions, and it will be hard to track down which definition is used.

Reply:
That was a great catch @ShimShtein 
updated it :)